framework/JobHandler.py

```diff
@@ -232,7 +262,112 @@ def __getLocalHost(self):
 
-  def __runRemoteListeningSockets(self,address):
+  def __shutdownParallel(self):
+    """
+      shutdown the parallel protocol
+      @ In, None
+      @ Out, None
+    """
+    if _rayAvail and self.rayServer is not None and not self.rayInstanciatedOutside:
+      # we need to ssh and stop each remote node cluster (ray)
+      servers = []
+      if 'remoteNodes' in self.runInfoDict:
+        servers += self.runInfoDict['remoteNodes']
+      if 'headNode' in self.runInfoDict:
+        servers += [self.runInfoDict['headNode']]
+      # get local enviroment
+      localenv = os.environ.copy()
+      localenv["PYTHONPATH"] = os.pathsep.join(sys.path)
+      for nodeAddress in servers:
+        self.raiseAMessage("Shutting down ray at address: "+ nodeAddress)
+        command="ray stop"
+        rayTerminate = utils.pickleSafeSubprocessPopen(['ssh',nodeAddress.split(":")[0],"COMMAND='"+command+"'",self.runInfoDict['RemoteRunCommand']],shell=False,env=localenv)
+        rayTerminate.wait()
+        if rayTerminate.returncode != 0:
+          self.raiseAWarning("RAY FAILED TO TERMINATE ON NODE: "+nodeAddress)
+      # shutdown ray API (object storage, plasma, etc.)
+      ray.shutdown()
+
+  def __runHeadNode(self, nProcs):
+    """
+      Method to activate the head ray server
+      @ In, nProcs, int, the number of processors
+      @ Out, address, str, the retrieved address (ip:port)
+      @ Out, redisPassword, str, the redis password
+    """
+    address, redisPassword = None, None
+    # get local enviroment
+    localenv = os.environ.copy()
+    localenv["PYTHONPATH"] = os.pathsep.join(sys.path)
+    if _rayAvail:
+      command = ["ray","start","--head"]
+      if nProcs is not None:
+        command.append("--num-cpus="+str(nProcs))
+      outFile = open("ray_head.ip", 'w')
+      rayStart = utils.pickleSafeSubprocessPopen(command,shell=False,stdout=outFile, stderr=outFile, env=localenv)
+      rayStart.wait()
+      outFile.close()
+      if rayStart.returncode != 0:
+        self.raiseAnError(RuntimeError, "RAY failed to start on the --head node! Return code is {}".format(rayStart.returncode))
+      else:
+        address, redisPassword = self.__getRayInfoFromStart("ray_head.ip")
+    return address, redisPassword
+
+  def __getRayInfoFromStart(self, rayLog):
+    """
+      Read Ray info from shell return script
+      @ In, rayLog, str, the ray output log
+      @ Out, address, str, the retrieved address (ip:port)
+      @ Out, redisPassword, str, the redis password
+    """
+    address, redisPassword = None, None
+    with open(rayLog, 'r') as rayLogObj:
+      for line in rayLogObj.readlines():
+        if "ray start" in line.strip():
+          ix = line.strip().find("ray start")
+          address, redisPassword = line.strip()[ix:].replace("ray start","").strip().split()
+          address = address.split("=")[-1].replace("'","")
+          redisPassword = redisPassword.split("=")[-1].replace("'","")
```

Comment:
Hm, it might be worth checking the other side of the split.  As in (and ditto for redisPassword):
```python
address_arg, address = address.replace("'","").split("=")
if address_arg != "--address":
  self.raiseAWarning("Unexpected ray start:" + line)
```

Reply:
added